### PR TITLE
Improve consistency of Shared and Permalink blocks UI.

### LINKED ISSUE #6469: Consider to improve consistency of the button labels and styles
Please see below, a quick comparison between the post permalink buttons and the shared block button. Initially, the both display an "Edit" button with the same style:

screen shot 2018-04-27 at 18 20 10

Instead, when editing:
- the first button becomes "OK" and is styled like a default button
- the second button becomes "Save" and is styled like a primary button (I understand it's to differentiate it from the "Cancel" button)

screen shot 2018-04-27 at 18 20 22

A few considerations:
- what is the standard text for a save action in Gutenberg? "OK" or "Save"? Seems to me using always the same text would improve consistency and make things easier for users
- should the "OK" button be styled as a primary button?
- what is the standard for a save button, should it always be paired with a "Cancel" button? 

/Cc @karmatosed

### PR DESCRIPTION
**Pull request by afercia**
This PR tries to improve consistency of the Post Permalink and Shared Block UIs, specifically about buttons and interaction.

- post permalink: changes "OK" to "Save"
- shared block:
  - removes primary style from the Save button
  - removes the "Cancel" button
  - closes the edit form when clicking outside of it, inspired by what [react-click-outside](https://github.com/kentor/react-click-outside/blob/e3e133799c51f896c1f65a799f0007827dd48248/index.js) does
  - (for keyboard users, it already closes when pressing Escape)

Screenshot:

screen shot 2018-04-28 at 14 11 17

Fixes #6469

### CODE REVIEW

**Review by karmatosed — APPROVED** [review]
Looks good to me and thanks for the speed here.

File: core-blocks/block/edit-panel/index.js [hunk 1]
```diff
@@ -22,9 +22,29 @@ class SharedBlockEditPanel extends Component {
...
+	componentDidMount() {
+		document.addEventListener( 'click', this.handleClickOutside, true );
+	}
+
+	componentWillUnmount() {
+		document.removeEventListener( 'click', this.handleClickOutside, true );
+	}
+
+	handleClickOutside( event ) {
+		if ( ! this.editForm.current ) {
```
Comment (aduth) [inline core-blocks/block/edit-panel/index.js]: When will this be the case?

File: core-blocks/block/edit-panel/index.js [hunk 2]
```diff
@@ -22,9 +22,29 @@ class SharedBlockEditPanel extends Component {
...
 		this.editButton = createRef();
+		this.editForm = createRef();
 		this.handleFormSubmit = this.handleFormSubmit.bind( this );
 		this.handleTitleChange = this.handleTitleChange.bind( this );
 		this.handleTitleKeyDown = this.handleTitleKeyDown.bind( this );
+		this.handleClickOutside = this.handleClickOutside.bind( this );
+	}
+
+	componentDidMount() {
+		document.addEventListener( 'click', this.handleClickOutside, true );
```
Comment (aduth) [inline core-blocks/block/edit-panel/index.js]: I'd have considered suggesting [`withGlobalEvents`](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-global-events), though it doesn't support `useCapture`. Out of curiosity, why was that needed here?

**Review by aduth — CHANGES_REQUESTED** [review]
This makes it impossible to edit the contents of the shared block, since editing is canceled when clicking outside the panel (into the block contents itself).

**Comment by karmatosed** [comment]
@chrisvanpatten lets maybe check other labels and go with consistency. Either way lets use that as a pattern @afercia.

File: core-blocks/block/edit-panel/index.js (hunk 1 above)
Comment (afercia) [reply core-blocks/block/edit-panel/index.js]: The edit form is a separate component and mounted only when editing, If you "click outside" when _not_ editing, then the returned DOM element is `null`.

File: core-blocks/block/edit-panel/index.js (hunk 2 above)
Comment (afercia) [reply core-blocks/block/edit-panel/index.js]: From the `react-click-outside` (which inspired this bit of code) readme: https://github.com/kentor/react-click-outside#details

> The enhanceWithClickOutside function wraps the provided component in another component that registers a click handler on document for the event capturing phase. Using the event capturing phase prevents elements with a click handler that calls stopPropagation from cancelling the click event that would eventually trigger the component's handleClickOutside function.

File: core-blocks/block/edit-panel/index.js (hunk 1 above)
Comment (aduth) [reply core-blocks/block/edit-panel/index.js]: But isn't the callback only bound to the `click` event when it is in-fact mounted?

**Comment by afercia** [comment]
@aduth I've refactored that check, to make it more clear.

**Comment by afercia** [comment]
> This makes it impossible to edit the contents of the shared block, since editing is canceled when clicking outside the panel

Ah, interesting. I've missed that the "Edit" button, which is visually associated to the _shared block name_ also gives access to editing the _contents_ of the shared block. A few considerations:
- at this point, I'm not so sure that when clicking "Edit" focus should be moved to the name field, as that's not the only editable thing in the block and not necessarily the primary action
- maybe the "Edit" button should be placed elsewhere, to make clear the whole block is editable
- the "click outside" behavior doesn't make sense, I'd say to remove it and close after saving is successful 
/Cc @aduth @karmatosed

File: editor/components/post-permalink/style.scss [hunk 3]
```diff
@@ -29,6 +29,7 @@
 
 .editor-post-permalink__label {
 	margin: 0 10px 0 5px;
+	font-weight: 600;
```
Comment (aduth) [inline editor/components/post-permalink/style.scss]: Minor: Could / should we add `<strong>` or `<b>` as a nested element of the label instead?

**Comment by karmatosed** [comment]
> Yep, this seems reasonable.

Just adding as pinged that I also agree.

File: editor/components/post-permalink/style.scss (hunk 3 above)
Comment (afercia) [reply editor/components/post-permalink/style.scss]: Not sure. `<strong>` adds a semantic meaning of "strong emphasis" which we don't need here, this element is a label and doesn't need emphasis. `<b>` is only presentational, so at this point why not use just CSS?

Comment (SuperGeniusZeb) [reply editor/components/post-permalink/style.scss]: `<b>` actually does have semantic meaning as a "bring attention to" element.

>The `<b>` element [represents](https://www.w3.org/TR/html53/dom.html#represent) a span of text to which attention is being drawn for utilitarian purposes without conveying any extra importance and with no implication of an alternate voice or mood, such as key words in a document abstract, product names in a review, actionable words in interactive text-driven software, or an article lede.

https://www.w3.org/TR/html53/textlevel-semantics.html#the-b-element
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/b

Comment (afercia) [reply editor/components/post-permalink/style.scss]: > without conveying any extra importance .. etc.

> The b element should be used as a last resort when no other element is more appropriate. In particular, headings should use the h1 to h6 elements, stress emphasis should use the em element, importance should be denoted with the strong element, and text marked or highlighted should use the mark element.

To me, this means it's basically a presentational element to "highlight" some text. So why we should use an additional element? For presentation, CSS is more appropriate. Also, the current official recommendation is HTML 5.2. Instead. 5.3 is still a working draft.

**Comment by afercia** [comment]
I've removed the "click outside" behavior: the block name form now closes on Save.

As per moving focus when editing, I've tried to change it and still not fully convinced. I'd tend to think the root issue is in the placement of the "Edit" button.:

screen shot 2018-05-14 at 18 22 57

Since it's close to the name form, it's really not so clear that also the shared block content becomes editable. Focusing the name field, makes the previous content hard to get to for keyboard users, as they're forced to reverse-tab. Also, there's no visual indication that the block content is editable:

screen shot 2018-05-14 at 18 22 35

- I've tried to move focus to the editable content: this ended up in duplicating most of the code already used in `BlockListBlock` > `focusTabbable()` 
- other options could be:
  - place the "Edit" button in a different spot
  - use two different "Edit" buttons, one for the block content, the other one for the block name

I'd appreciate some feedback, any thoughts welcome.

**Comment by aduth** [comment]
I can't speak to the design / UX impact of separate edit buttons, but so far as shifting focus to the editable content: it seems we lack but could benefit from a consistent programmatic way to focus a block's editable content. Both the [`SharedBlockEdit`](https://github.com/WordPress/gutenberg/blob/7ad3fa44230a664228bfafba14516f70441d3d96/core-blocks/block/edit.js#L101-L108) and [`BlockListBlock`](https://github.com/WordPress/gutenberg/blob/7ad3fa44230a664228bfafba14516f70441d3d96/editor/components/block-list/block.js#L553-L564) components render a `BlockEdit` component. Maybe we should move `focusTabbable` from `BlockListBlock` into `BlockEdit`, then both `BlockListBlock` and `SharedBlockEdit` could call the `focusTabbable` function from a `ref` assigned to their rendered `BlockEdit`.

Let me know if this sounds reasonable. I could assist with these changes if needed.

**Comment by aduth** [comment]
_Responding in passing, will give a closer look as I catch up._ This may be where we need [`ref` forwarding](https://reactjs.org/docs/forwarding-refs.html) on our higher-order components (probably via the `createHigherOrderComponent` abstraction, cc @gziolo I think we may have mentioned this before?).

**Comment by gziolo** [comment]
Given that we usually compose HOCs it should be rather included in the existing helper method. Ideally, we should explore if we can detect if ref was passed and try to forward it. Do you think it would be a breaking change? Do you have any specific use case where you know it would change behavior? So far it was rather a limitation.

**Comment by nerrad** [comment]
I shouldn't have been so imperative in my statement.  Rather than "It'd be a fairly breaking change" I should have written "It might be a breaking change".  I'll explore some more and ya I might be able to just detect if ref is already passed.

**Comment by gziolo** [comment]
Cool, we have time until 3.3 ships (2 weeks?) for breaking changes. In case we need them 😃 

Another thing we still didn’t agree on is if we should move a few general HOCs and related helpers from element module to its own module. This was raised a few times because sometimes you need to pull in all components to just use HOC which is a more general concept, on the other hand composing HOCs isn’t really an essential part of element package 🤔

**Comment by afercia** [comment]
Rebased keeping the change from #7987. Screenshot:

screen shot 2018-08-08 at 14 29 07

Note 1:
unrelated but the permalink input field misses the new `border-radius` style recently added /Cc @jasmussen 

Note 2:
unrelated (happens also on master) the permalink UI needs some love in the responsive view, especially when the full URL is long. Screenshot from master:

screen shot 2018-08-08 at 14 47 41

**Comment by Soean** [comment]
I haven't seen this PR before... Most of the changed were merged in #7905 + #10563 missing from this PR. I opened a separate PR for the permalink label: #10566
